### PR TITLE
Feat/explicit-contract-version

### PR DESCRIPTION
**Pull request by jcnelson**
This addresses #3131 and #2719, as well as #2609.  It introduces a `SmartContract` payload variant that takes an explicit Clarity version, so that post-2.1 developers can require that the contract adhere to Clarity 1 or Clarity 2 rules.  If they do not do so, then the epoch default Clarity version is used (which will be Clarity 2 in 2.1).

The PR removes most uses of `ClarityVersion::default_for_epoch()` and replaces them with a transaction-supplied Clarity version.  When a transaction is processed, a new method `StacksChainState::get_tx_clarity_version()` is called to deduce which version of Clarity to supply the Clarity VM.  The rules are as follows:

* If the transaction is a smart contract transaction, and a version is explicitly given, then use that.
* If the transaction is a smart contract transaction, and a version is not given, then use the epoch default.
* If the transaction is a contract-call transaction, then load up the target contract and use its version.
* In all other cases, use the epoch default.

This PR tests #2719 by addressing #3131.  It introduces a "torture test" for cross-contract and cross-Clarity-version contract calls in `chainstate::stacks::tests::block_construction::test_contract_call_across_clarity_versions`.  It verifies the following:

* The method invoked by a contract call transaction runs with the Clarity version that the contract that defined it used.  So for example, a Clarity1 contract can call a Clarity2 contract function, and that Clarity2 function can do Clarity2-specific things.

* The method invoked by an explicit `contract-call?` follows the above rule as well.

* A trait defined in Clarity1 can be implemented by a Clarity2 contract.

* A trait defined in Clarity2 can be implemented by a Clarity1 contract.

* Dynamic dispatches to traits will run with the Clarity version of the trait implementation.  So for example, if a Clarity1 contract calls into a trait that, regardless of the trait definition's Clarity version, has a Clarity2 implementation, then the dispatch will run in Clarity2.

* The closure of an `(at-block)` will run with the Clarity version in which it is defined, like any other code statement.  Similarly, contract-calls within the at-block will run in the target method's contract's Clarity version.

The PR addresses #2609 by ensuring that a block will not be processed if it contains 2.1-only transactions, such as pay-to-alt-principal coinbases and explicit-version smart contracts.  This gating logic not only applies in block validation (which covers both the blockchain-processing and miner), but also in mempool acceptance.

### CODE REVIEW

File: src/chainstate/stacks/db/transactions.rs [hunk 1]
```diff
@@ -1139,6 +1143,46 @@ impl StacksChainState {
...
+                                return Err(Error::InvalidStacksTransaction(msg, false));
+                            }
+                            _ => {
+                                panic!("FATAL: contract analysis data did not report a clarity version for '{}'", &contract_id);
+                            }
+                        }
+                    }
+                };
+                clarity_version
+            }
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/chainstate/stacks/db/transactions.rs:1177]: For reference:

Contract publish invokes `Contract::initialize_from_ast` (https://github.com/stacks-network/stacks-blockchain/blob/next/clarity/src/vm/contracts.rs#L35), which sets the contract context (overriding the `default_contract`). Here, it's still necessary to get the clarity version "passed in", as you can tell from the link, it uses a default version.

Contract call invokes `Environment::execute_contract` (https://github.com/stacks-network/stacks-blockchain/blob/next/clarity/src/vm/contexts.rs#L1032), which sets the contract context (in `execute_function_as_transaction`). Here, no clarity version would need to be "passed in".

I *think* the best thing to do would be to actually get rid of the `default_contract`.

[the remark this answers is not in the document]
Comment (jcnelson) [reply src/chainstate/stacks/db/transactions.rs:1177]: > @kantai which is essentially encoding the "what clarity version should run?" decision twice, where one of the instances shouldn't ever matter

Can we just remove `default_contract`?

**Review by gregorycoppola — APPROVED** [review]
Overall, the algorithm described in the PR description sounds pretty clean.

Thanks for a large PR!

File: src/chainstate/stacks/db/transactions.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/chainstate/stacks/db/transactions.rs]: Sure, removed.

File: src/chainstate/stacks/tests/block_construction.rs [hunk 2]
```diff
@@ -3189,3 +3189,560 @@ fn test_build_microblock_stream_forks_with_descendants() {
...
+            end_height: 1, // NOTE: the first 25 burnchain blocks have no sortition
+            block_limit: ExecutionCost::max_value(),
+            network_epoch: PEER_VERSION_EPOCH_2_0,
+        },
+        StacksEpoch {
+            epoch_id: StacksEpochId::Epoch2_05,
+            start_height: 1,
+            end_height: 2, // NOTE: the first 25 burnchain blocks have no sortition
+            block_limit: ExecutionCost::max_value(),
+            network_epoch: PEER_VERSION_EPOCH_2_0,
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/chainstate/stacks/tests/block_construction.rs:3247]: Yup

File: src/clarity_vm/tests/large_contract.rs [hunk 3]
```diff
@@ -668,8 +691,8 @@ pub fn let_memory_test() {
 
         contract.push_str(") 1)");
 
-        conn.as_transaction(|conn| {
-            let (ct_ast, _ct_analysis) = conn
+        conn.as_transaction(clarity_version, |conn| {
+            let (ct_ast, ct_analysis) = conn
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/clarity_vm/tests/large_contract.rs]: At one point in this PR's development, they were used.  But I just reverted this.

**Review by obycode — APPROVED** [review]
LGTM

**Pull request merged**